Script providers not working on OpenJDK 17 due to Nashorn being deprecated

### Describe the bug

Hitting this below exception when I run the `org.keycloak.testsuite.authz.*` module tests 

and here is the reference for the deprecation notice in openjdk - https://openjdk.java.net/jeps/372

I was able to confirm the same runs just fine with JDK11 runtime for the authserver. This would be a blocker for the Java 17 certification tests.

```Caused by: java.lang.IllegalStateException: Could not find ScriptEngine for script: Script{id='null', realmId='0cfba609-5d40-404f-a3eb-769cf46dcff8', name='Grant Policy', type='text/javascript', code='$evaluation.grant();', description='null'}```

This is also was previously discussed in [KEYCLOAK-12755](https://issues.redhat.com/browse/KEYCLOAK-12755)

### Version

17.0.0-SNAPSHOT, OpenJDK17

### Expected behavior

We would like the javascript engine to initialize and not fail on the authorization work flows

### Actual behavior

The suspicion is that because of the deprecated Nashorn js engine in jdk17, we are getting a `Could not find ScriptEngine for script` IllegalStateExceptions.

I think we also want a proper exception to bubble up the stack trace, as currently we get a generic exception which is 
`Unexpected error while evaluating permissions: java.lang.RuntimeException: Failed to evaluate permissions` even though we get the actual problem the stack trace down the line.

### How to Reproduce?

run the authz module testsuite using the below mvn command

set the right binary path for the JDK17_HOME and MVN_SETTINGS_PATH before you run it.

```
mvn -f testsuite/integration-arquillian/tests/base/pom.xml clean install -Dauth.server.java.home=${JDK17_HOME} -Dauth.server.memory.settings="-Xms128m -Xmx512m -XX:MetaspaceSize=96m -XX:MaxMetaspaceSize=256m" -Dapp.server.memory.settings="-Xms128m -Xmx512m -XX:MetaspaceSize=96m -XX:MaxMetaspaceSize=256m" -B -s ${MVN_SETTINGS_PATH} -Dsettings.path=${MVN_SETTINGS_PATH} -Dsurefire.memory.Xms=512m -Dsurefire.memory.Xmx=1536m -Dinsecure.repositories=WARN -Pauth-server-wildfly -Pjava11-auth-server -Dtest=org.keycloak.testsuite.authz.AuthorizationTest

```

### Anything else?

Issue full stack trace:

```
2022-02-01 20:07:38,427 ERROR [org.keycloak.authorization.authorization.AuthorizationTokenService] (default task-1) Unexpected error while evaluating permissions: java.lang.RuntimeException: Failed to evaluate permissions
        at org.keycloak.keycloak-server-spi-private@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.evaluation.DecisionPermissionCollector.onError(DecisionPermissionCollector.java:182)
        at org.keycloak.keycloak-server-spi-private@17.0.0-SNAPSHOT//org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator.evaluate(IterablePermissionEvaluator.java:71)
        at org.keycloak.keycloak-server-spi-private@17.0.0-SNAPSHOT//org.keycloak.authorization.permission.evaluator.IterablePermissionEvaluator.evaluate(IterablePermissionEvaluator.java:87)
        at org.keycloak.keycloak-services@17.0.0-SNAPSHOT//org.keycloak.authorization.authorization.AuthorizationTokenService.evaluatePermissions(AuthorizationTokenService.java:285)
        at org.keycloak.keycloak-services@17.0.0-SNAPSHOT//org.keycloak.authorization.authorization.AuthorizationTokenService.authorize(AuthorizationTokenService.java:222)
.
.
.
.
Caused by: java.lang.IllegalStateException: Could not find ScriptEngine for script: Script{id='null', realmId='0cfba609-5d40-404f-a3eb-769cf46dcff8', name='Grant Policy', type='text/javascript', code='$evaluation.grant();', description='null'}
        at org.keycloak.keycloak-services@17.0.0-SNAPSHOT//org.keycloak.scripting.DefaultScriptingProvider.createPreparedScriptEngine(DefaultScriptingProvider.java:106)
        at org.keycloak.keycloak-services@17.0.0-SNAPSHOT//org.keycloak.scripting.DefaultScriptingProvider.prepareEvaluatableScript(DefaultScriptingProvider.java:72)
        at org.keycloak.keycloak-services@17.0.0-SNAPSHOT//org.keycloak.scripting.DefaultScriptingProvider.prepareEvaluatableScript(DefaultScriptingProvider.java:33)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.js.JSPolicyProviderFactory.lambda$getEvaluatableScript$0(JSPolicyProviderFactory.java:109)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.js.ScriptCache.lambda$computeIfAbsent$0(ScriptCache.java:80)
        at java.base/java.util.HashMap.computeIfAbsent(HashMap.java:1220)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.js.ScriptCache.computeIfAbsent(ScriptCache.java:80)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.js.JSPolicyProviderFactory.getEvaluatableScript(JSPolicyProviderFactory.java:106)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.js.JSPolicyProvider.evaluate(JSPolicyProvider.java:46)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.permission.AbstractPermissionProvider.evaluate(AbstractPermissionProvider.java:56)
        at org.keycloak.keycloak-authz-policy-common@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.provider.permission.ResourcePolicyProvider.evaluate(ResourcePolicyProvider.java:47)
        at org.keycloak.keycloak-server-spi-private@17.0.0-SNAPSHOT//org.keycloak.authorization.policy.evaluation.DefaultPolicyEvaluator.lambda$createPolicyEvaluator$0(DefaultPolicyEvaluator.java:116)

```

--- a/model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java
+++ b/model/jpa/src/main/java/org/keycloak/connections/jpa/util/JpaUtils.java
@@ -25,7 +25,7 @@ import org.hibernate.jpa.boot.internal.ParsedPersistenceXmlDescriptor;
 import org.hibernate.jpa.boot.internal.PersistenceXmlParser;
 import org.hibernate.jpa.boot.spi.Bootstrap;
 import org.keycloak.connections.jpa.entityprovider.JpaEntityProvider;
-import org.keycloak.connections.jpa.entityprovider.ProxyClassLoader;
+import org.keycloak.utils.ProxyClassLoader;
 import org.keycloak.models.KeycloakSession;
 
 import javax.persistence.EntityManager;

--- a/pom.xml
+++ b/pom.xml
@@ -118,6 +118,7 @@
         <xmlsec.version>2.2.3</xmlsec.version>
         <glassfish.json.version>1.1.6</glassfish.json.version>
         <wildfly.common.version>1.6.0.Final</wildfly.common.version>
+        <nashorn.version>15.3</nashorn.version>
         <ua-parser.version>1.5.2</ua-parser.version>
         <picketbox.version>5.0.3.Final</picketbox.version>
         <google.guava.version>30.1-jre</google.guava.version>
@@ -590,6 +591,11 @@
                 <groupId>org.glassfish</groupId>
                 <artifactId>jakarta.json</artifactId>
                 <version>${glassfish.json.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.openjdk.nashorn</groupId>
+                <artifactId>nashorn-core</artifactId>
+                <version>${nashorn.version}</version>
             </dependency>
 
             <!-- Twitter -->

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusPlatform.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusPlatform.java
@@ -19,13 +19,20 @@ package org.keycloak.quarkus.runtime.integration;
 
 import java.io.File;
 import java.io.IOException;
+import java.net.MalformedURLException;
+import java.net.URL;
+import java.net.URLClassLoader;
 import java.nio.file.Files;
 import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.LinkedList;
 import java.util.List;
 import java.util.concurrent.CopyOnWriteArrayList;
 import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.stream.Stream;
 
 import org.jboss.logging.Logger;
+import org.keycloak.Config;
 import org.keycloak.platform.Platform;
 import org.keycloak.platform.PlatformProvider;
 import org.keycloak.quarkus.runtime.InitializationException;
@@ -158,5 +165,10 @@ public class QuarkusPlatform implements PlatformProvider {
 
     private void reset() {
         deferredExceptions.clear();
+    }
+
+    @Override
+    public ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig) {
+        return null;
     }
 }

--- a/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusPlatform.java
+++ b/quarkus/runtime/src/main/java/org/keycloak/quarkus/runtime/integration/QuarkusPlatform.java
@@ -169,6 +169,7 @@ public class QuarkusPlatform implements PlatformProvider {
 
     @Override
     public ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig) {
+        // It is fine to return null assuming that nashorn and it's dependencies are included on the classpath (usually "providers" directory)
         return null;
     }
 }

--- a/server-spi-private/src/main/java/org/keycloak/utils/ProxyClassLoader.java
+++ b/server-spi-private/src/main/java/org/keycloak/utils/ProxyClassLoader.java
@@ -1,5 +1,5 @@
 /*
- * Copyright 2016 Red Hat, Inc. and/or its affiliates
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
  * and other contributors as indicated by the @author tags.
  *
  * Licensed under the Apache License, Version 2.0 (the "License");
@@ -13,13 +13,18 @@
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
+ *
  */
 
-package org.keycloak.connections.jpa.entityprovider;
+package org.keycloak.utils;
 
+import java.io.IOException;
 import java.net.URL;
+import java.util.Arrays;
 import java.util.Collection;
-import java.util.HashSet;
+import java.util.Enumeration;
+import java.util.Iterator;
+import java.util.LinkedHashSet;
 import java.util.Set;
 
 /**
@@ -29,7 +34,8 @@ import java.util.Set;
  * Effectively it forms a proxy to one or more other classloaders.
  * 
  * The way it works:
- * - Get all (unique) classloaders from all provided classes
+ * - Get list of classloaders, which will be used as "delegates" when loaded classes or resources.
+ *   - Can be retrived from provided classloaders or alternatively from the provided classes where the "delegate classloaders" will be determined from the classloaders of given classes
  * - For each class or resource that is 'requested':
  *   - First try all provided classloaders and if we have a match, return that
  *   - If no match was found: proceed with 'normal' classloading in 'current classpath' scope
@@ -41,17 +47,28 @@ public class ProxyClassLoader extends ClassLoader {
 
     private Set<ClassLoader> classloaders;
 
-    public ProxyClassLoader(Collection<Class<?>> classes, ClassLoader parentClassLoader) {
-    	super(parentClassLoader);
-    	init(classes);
+    /**
+     * Init classloader with the list of given delegates
+     * @param delegateClassLoaders
+     */
+    public ProxyClassLoader(ClassLoader... delegateClassLoaders) {
+        if (delegateClassLoaders == null || delegateClassLoaders.length == 0) {
+            throw new IllegalStateException("At least one classloader to delegate must be provided");
+        }
+        classloaders = new LinkedHashSet<>();
+        classloaders.addAll(Arrays.asList(delegateClassLoaders));
     }
-    
+
+    /**
+     * Get all unique classloaders from the provided classes to be used as "Delegate classloaders"
+     * @param classes
+     */
     public ProxyClassLoader(Collection<Class<?>> classes) {
     	init(classes);
     }
 
     private void init(Collection<Class<?>> classes) {
-        classloaders = new HashSet<>();
+        classloaders = new LinkedHashSet<>();
         for (Class<?> clazz : classes) {
             classloaders.add(clazz.getClassLoader());
         }
@@ -84,4 +101,33 @@ public class ProxyClassLoader extends ClassLoader {
         return super.getResource(name);
     }
 
+    @Override
+    public Enumeration<URL> getResources(String name) throws IOException {
+        final LinkedHashSet<URL> resourceUrls = new LinkedHashSet();
+
+        for (ClassLoader classloader : classloaders) {
+            Enumeration<URL> child = classloader.getResources(name);
+
+            while (child.hasMoreElements()) {
+                resourceUrls.add(child.nextElement());
+            }
+        }
+
+        return new Enumeration<URL>() {
+            final Iterator<URL> resourceUrlIterator = resourceUrls.iterator();
+
+            public boolean hasMoreElements() {
+                return this.resourceUrlIterator.hasNext();
+            }
+
+            public URL nextElement() {
+                return (URL)this.resourceUrlIterator.next();
+            }
+        };
+    }
+
+    @Override
+    public String toString() {
+        return "ProxyClassLoader: Delegates: " + classloaders;
+    }
 }

--- a/services/src/main/java/org/keycloak/platform/PlatformProvider.java
+++ b/services/src/main/java/org/keycloak/platform/PlatformProvider.java
@@ -19,6 +19,8 @@ package org.keycloak.platform;
 
 import java.io.File;
 
+import org.keycloak.Config;
+
 public interface PlatformProvider {
     
     void onStartup(Runnable runnable);
@@ -32,5 +34,18 @@ public interface PlatformProvider {
      * The directory should be usually inside the corresponding server directory. In production, it should not be system directory like "/tmp" .
      */
     File getTmpDirectory();
+
+
+    /**
+     * Returns classloader to load script engine. Classloader should contain the implementation of {@link javax.script.ScriptEngineFactory}
+     * and it's definition inside META-INF/services of the jar file(s), which will be provided by this classloader.
+     *
+     * This method can return null and in that case, the default Keycloak services classloader will be used for load script engine. Note that java versions earlier than 15 always contain
+     * the "nashorn" script engine by default on the classpath (it is part of the Java platform itself) and hence for them it is always fine to return null (unless you want to override default engine)
+     *
+     * @param scriptProviderConfig Configuration scope of the "default" provider of "scripting" SPI. It can contain some config properties for the classloader (EG. file path)
+     * @return classloader or null
+     */
+    ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig);
 
 }

--- a/services/src/main/java/org/keycloak/scripting/DefaultScriptingProvider.java
+++ b/services/src/main/java/org/keycloak/scripting/DefaultScriptingProvider.java
@@ -24,7 +24,9 @@ import javax.script.ScriptEngineManager;
 import javax.script.ScriptException;
 
 import org.keycloak.models.ScriptModel;
+import org.keycloak.platform.Platform;
 import org.keycloak.services.ServicesLogger;
+import org.keycloak.utils.ProxyClassLoader;
 
 /**
  * A {@link ScriptingProvider} that uses a {@link ScriptEngineManager} to evaluate scripts with a {@link ScriptEngine}.
@@ -125,8 +127,17 @@ public class DefaultScriptingProvider implements ScriptingProvider {
     private ScriptEngine lookupScriptEngineFor(ScriptModel script) {
         ClassLoader cl = Thread.currentThread().getContextClassLoader();
         try {
-            Thread.currentThread().setContextClassLoader(DefaultScriptingProvider.class.getClassLoader());
-            return factory.getScriptEngineManager().getEngineByMimeType(script.getMimeType());
+            ClassLoader scriptClassLoader = Platform.getPlatform().getScriptEngineClassLoader(factory.getConfig());
+
+            // Also need to use classloader of keycloak services itself to be able to use keycloak classes in the scripts
+            if (scriptClassLoader != null) {
+                scriptClassLoader = new ProxyClassLoader(scriptClassLoader, DefaultScriptingProvider.class.getClassLoader());
+            } else {
+                scriptClassLoader = DefaultScriptingProvider.class.getClassLoader();
+            }
+
+            Thread.currentThread().setContextClassLoader(scriptClassLoader);
+            return new ScriptEngineManager().getEngineByMimeType(script.getMimeType());
         }
         finally {
             Thread.currentThread().setContextClassLoader(cl);

--- a/services/src/main/java/org/keycloak/scripting/DefaultScriptingProviderFactory.java
+++ b/services/src/main/java/org/keycloak/scripting/DefaultScriptingProviderFactory.java
@@ -23,10 +23,8 @@ import org.jboss.logging.Logger;
 import org.keycloak.Config;
 import org.keycloak.models.KeycloakSession;
 import org.keycloak.models.KeycloakSessionFactory;
-import org.keycloak.protocol.oidc.OIDCWellKnownProviderFactory;
 
 import javax.script.ScriptEngine;
-import javax.script.ScriptEngineManager;
 
 /**
  * @author <a href="mailto:thomas.darimont@gmail.com">Thomas Darimont</a>
@@ -35,30 +33,28 @@ public class DefaultScriptingProviderFactory implements ScriptingProviderFactory
 
     private static final Logger logger = Logger.getLogger(DefaultScriptingProviderFactory.class);
 
-    static final String ID = "script-based-auth";
-
-    private ScriptEngineManager scriptEngineManager;
+    static final String ID = "default";
 
     private boolean enableScriptEngineCache;
 
     // Key is mime-type. Value is engine for the particular mime-type. Cache can be used when the scriptEngine can be shared across multiple threads / requests (which is the case for nashorn)
     private Map<String, ScriptEngine> scriptEngineCache;
 
+    private Config.Scope config;
+
     @Override
     public ScriptingProvider create(KeycloakSession session) {
-        lazyInit();
-
         return new DefaultScriptingProvider(this);
     }
 
     @Override
     public void init(Config.Scope config) {
+        this.config = config;
         this.enableScriptEngineCache = config.getBoolean("enable-script-engine-cache", true);
         logger.debugf("Enable script engine cache: %b", this.enableScriptEngineCache);
-    }
-
-    ScriptEngineManager getScriptEngineManager() {
-        return scriptEngineManager;
+        if (enableScriptEngineCache) {
+            scriptEngineCache = new ConcurrentHashMap<>();
+        }
     }
 
     boolean isEnableScriptEngineCache() {
@@ -67,6 +63,10 @@ public class DefaultScriptingProviderFactory implements ScriptingProviderFactory
 
     Map<String, ScriptEngine> getScriptEngineCache() {
         return scriptEngineCache;
+    }
+
+    Config.Scope getConfig() {
+        return config;
     }
 
     @Override
@@ -82,19 +82,6 @@ public class DefaultScriptingProviderFactory implements ScriptingProviderFactory
     @Override
     public String getId() {
         return ID;
-    }
-
-    private void lazyInit() {
-        if (scriptEngineManager == null) {
-            synchronized (this) {
-                if (scriptEngineManager == null) {
-                    scriptEngineManager = new ScriptEngineManager();
-                    if (enableScriptEngineCache) {
-                        scriptEngineCache = new ConcurrentHashMap<>();
-                    }
-                }
-            }
-        }
     }
 
 }

--- a/testsuite/integration-arquillian/servers/auth-server/jboss/common/ant/configure.xml
+++ b/testsuite/integration-arquillian/servers/auth-server/jboss/common/ant/configure.xml
@@ -257,4 +257,18 @@
             </filterset>
         </copy>
     </target>
+
+    <!-- Needed on Java 15 and later -->
+    <target name="deploy-nashorn-module">
+        <copy todir="${cli.tmp.dir}">
+            <resources>
+                <file file="${common.resources}/jboss-cli/deploy-nashorn-module.cli"/>
+            </resources>
+            <filterset>
+                <filter token="NASHORN_JAR" value="${project.build.directory}/nashorn/nashorn-core-${nashorn.version}.jar"/>
+            </filterset>
+        </copy>
+        <echo>Nashorn module deployed</echo>
+    </target>
+
 </project>

--- a/testsuite/integration-arquillian/servers/auth-server/jboss/common/jboss-cli/deploy-nashorn-module.cli
+++ b/testsuite/integration-arquillian/servers/auth-server/jboss/common/jboss-cli/deploy-nashorn-module.cli
@@ -1,0 +1,5 @@
+echo *** Installing nashorn-core module ***
+module add --module-root-dir=../modules/system/layers/keycloak/ \
+           --name=org.openjdk.nashorn.nashorn-core \
+           --resources=@NASHORN_JAR@ \
+           --dependencies=asm.asm,jdk.dynalink

--- a/testsuite/integration-arquillian/servers/auth-server/jboss/pom.xml
+++ b/testsuite/integration-arquillian/servers/auth-server/jboss/pom.xml
@@ -732,5 +732,62 @@
             </properties>
         </profile>
 
+        <!-- Nashorn script engine needs to be manually added for the new Java versions as it is not part of the JDK anymore  -->
+        <profile>
+            <id>jdk15</id>
+            <activation>
+                <jdk>[15,)</jdk>
+            </activation>
+
+            <build>
+                <pluginManagement>
+                    <plugins>
+                        <plugin>
+                            <artifactId>maven-dependency-plugin</artifactId>
+                            <executions>
+                                <execution>
+                                    <id>copy-nashorn-module</id>
+                                    <phase>generate-resources</phase>
+                                    <goals>
+                                        <goal>copy</goal>
+                                    </goals>
+                                    <configuration>
+                                        <artifactItems>
+                                            <artifactItem>
+                                                <groupId>org.openjdk.nashorn</groupId>
+                                                <artifactId>nashorn-core</artifactId>
+                                                <version>${nashorn.version}</version>
+                                                <type>jar</type>
+                                            </artifactItem>
+                                        </artifactItems>
+                                        <outputDirectory>${project.build.directory}/nashorn</outputDirectory>
+                                        <overWriteIfNewer>true</overWriteIfNewer>
+                                    </configuration>
+                                </execution>
+                            </executions>
+                        </plugin>
+                        <plugin>
+                            <groupId>org.apache.maven.plugins</groupId>
+                            <artifactId>maven-antrun-plugin</artifactId>
+                            <executions>
+                                <execution>
+                                    <id>deploy-nashorn-module</id>
+                                    <phase>generate-resources</phase>
+                                    <goals>
+                                        <goal>run</goal>
+                                    </goals>
+                                    <configuration>
+                                        <target>
+                                            <ant antfile="${common.resources}/ant/configure.xml" target="deploy-nashorn-module" />
+                                        </target>
+                                    </configuration>
+                                </execution>
+                            </executions>
+                        </plugin>
+                    </plugins>
+                </pluginManagement>
+            </build>
+        </profile>
+
     </profiles>
 </project>

--- a/testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
+++ b/testsuite/integration-arquillian/servers/auth-server/quarkus/pom.xml
@@ -262,5 +262,43 @@
                 <auth.server.quarkus.cluster.config>ha</auth.server.quarkus.cluster.config>
             </properties>
         </profile>
+
+        <!-- Nashorn script engine needs to be manually added for the new Java versions as it is not part of the JDK anymore  -->
+        <profile>
+            <id>jdk15</id>
+            <activation>
+                <jdk>[15,)</jdk>
+            </activation>
+
+            <dependencies>
+                <dependency>
+                    <groupId>org.openjdk.nashorn</groupId>
+                    <artifactId>nashorn-core</artifactId>
+                </dependency>
+            </dependencies>
+
+            <build>
+                <plugins>
+                    <plugin>
+                        <groupId>org.apache.maven.plugins</groupId>
+                        <artifactId>maven-dependency-plugin</artifactId>
+                        <executions>
+                            <execution>
+                                <id>copy-dependencies-quarkus</id>
+                                <phase>generate-resources</phase>
+                                <goals>
+                                    <goal>copy-dependencies</goal>
+                                </goals>
+                                <configuration>
+                                    <outputDirectory>${auth.server.home}/providers</outputDirectory>
+                                    <includeArtifactIds>nashorn-core,asm,asm-util,asm-commons</includeArtifactIds>
+                                </configuration>
+                            </execution>
+                        </executions>
+                    </plugin>
+                </plugins>
+            </build>
+        </profile>
+
     </profiles>
 </project>

--- a/testsuite/utils/pom.xml
+++ b/testsuite/utils/pom.xml
@@ -254,6 +254,12 @@
             <artifactId>mssql-jdbc</artifactId>
             <version>${mssql.driver.version}</version>
         </dependency>
+
+        <!-- Nashorn -->
+        <dependency>
+            <groupId>org.openjdk.nashorn</groupId>
+            <artifactId>nashorn-core</artifactId>
+        </dependency>
     </dependencies>
 
 

--- a/testsuite/utils/src/main/java/org/keycloak/testsuite/TestPlatform.java
+++ b/testsuite/utils/src/main/java/org/keycloak/testsuite/TestPlatform.java
@@ -22,6 +22,7 @@ import java.io.IOException;
 import java.nio.file.Files;
 
 import org.jboss.logging.Logger;
+import org.keycloak.Config;
 import org.keycloak.platform.PlatformProvider;
 
 public class TestPlatform implements PlatformProvider {
@@ -69,5 +70,11 @@ public class TestPlatform implements PlatformProvider {
             }
         }
         return tmpDir;
+    }
+
+    @Override
+    public ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig) {
+        // It is fine to return null as nashorn should be automatically included on the classpath of testsuite utils
+        return null;
     }
 }

--- a/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyPlatform.java
+++ b/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyPlatform.java
@@ -98,8 +98,7 @@ public class WildflyPlatform implements PlatformProvider {
             if (WildflyUtil.getJavaVersion() >= 15) {
                 log.warnf("Cannot find script engine in the JBoss module '%s'. Please add JavaScript engine to the specified JBoss Module or make sure it is available on the classpath", engineModule);
             } else {
-                // TODO:mposolda debug
-                log.infof("Cannot find script engine in the JBoss module '%s'. Will fallback to the default script engine", engineModule);
+                log.debugf("Cannot find script engine in the JBoss module '%s'. Will fallback to the default script engine", engineModule);
             }
             return null;
         }

--- a/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyPlatform.java
+++ b/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyPlatform.java
@@ -20,12 +20,20 @@ package org.keycloak.provider.wildfly;
 import java.io.File;
 
 import org.jboss.logging.Logger;
+import org.jboss.modules.Module;
+import org.jboss.modules.ModuleIdentifier;
+import org.jboss.modules.ModuleLoadException;
+import org.keycloak.Config;
+import org.keycloak.common.util.Environment;
 import org.keycloak.platform.PlatformProvider;
 import org.keycloak.services.ServicesLogger;
 
 public class WildflyPlatform implements PlatformProvider {
 
     private static final Logger log = Logger.getLogger(WildflyPlatform.class);
+
+    // In this module, the attempt to load script engine will be done by default
+    private static final String DEFAULT_SCRIPT_ENGINE_MODULE = "org.openjdk.nashorn.nashorn-core";
 
     Runnable shutdownHook;
 
@@ -73,5 +81,27 @@ public class WildflyPlatform implements PlatformProvider {
             }
         }
         return tmpDir;
+    }
+
+    @Override
+    public ClassLoader getScriptEngineClassLoader(Config.Scope scriptProviderConfig) {
+        String engineModule = scriptProviderConfig.get("script-engine-module");
+        if (engineModule == null) {
+            engineModule = DEFAULT_SCRIPT_ENGINE_MODULE;
+        }
+
+        try {
+            Module module = Module.getContextModuleLoader().loadModule(ModuleIdentifier.fromString(engineModule));
+            log.infof("Found script engine module '%s'", engineModule);
+            return module.getClassLoader();
+        } catch (ModuleLoadException mle) {
+            if (WildflyUtil.getJavaVersion() >= 15) {
+                log.warnf("Cannot find script engine in the JBoss module '%s'. Please add JavaScript engine to the specified JBoss Module or make sure it is available on the classpath", engineModule);
+            } else {
+                // TODO:mposolda debug
+                log.infof("Cannot find script engine in the JBoss module '%s'. Will fallback to the default script engine", engineModule);
+            }
+            return null;
+        }
     }
 }

--- a/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyUtil.java
+++ b/wildfly/extensions/src/main/java/org/keycloak/provider/wildfly/WildflyUtil.java
@@ -1,0 +1,40 @@
+/*
+ * Copyright 2022 Red Hat, Inc. and/or its affiliates
+ * and other contributors as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package org.keycloak.provider.wildfly;
+
+/**
+ * @author <a href="mailto:mposolda@redhat.com">Marek Posolda</a>
+ */
+public class WildflyUtil {
+
+    /**
+     * @return major version of java runtime like 8, 11 or 17
+     */
+    public static int getJavaVersion() {
+        String version = System.getProperty("java.version");
+        if (version.startsWith("1.")) {
+            version = version.substring(2, 3);
+        } else {
+            int dot = version.indexOf(".");
+            if (dot != -1) version = version.substring(0, dot);
+        }
+
+        return Integer.parseInt(version);
+    }
+}

